buildPython*: unconditionally add .override

sternenseemann opened the pull request:
Preserve laziness of `buildPython*` build-helper functions, by adding `override` unconditionally instead of via `optionalAttrs (f ? override)`.

Checking if `f` contains `override` causes it to be strictly evaluated. The check is also uneccessary because we know `makeOverridablePythonPackage` is always supplied a `callPackage`-result (either `callPackage ./mk-python-derivation.nix { }` or the python2 equivalent), so `override` should always be present in `f`.

This should resolve issues like `pkgsCross.ghcjs.buildPackages.nixpkgs-openjdk-updater` on staging right now (ff5098e3056110b7d40e49151248f90e43a6c7e1).

<details><summary>Original description</summary>
<p>


This change fixes two kinds of issues:

1. Fixes evaluation if `result` is neither a function nor an attribute set (which the code at least allowed before #366593).
2. Fix evaluation of buildPython* based packages in certain splicing situations by being lazier. An example of this is pkgsCross.ghcjs.buildPackages.nixpkgs-openjdk-updater on staging right now (ff5098e3056110b7d40e49151248f90e43a6c7e1). Apparently, a change that is not part of master, caused this to manifest on staging.

   By moving the addition of the set with .override inwards, we are able generating the outer __spliced set without forcing result, so we can actually access splices that don't throw.


</p>
</details> 

cc @wolfgangwalther @alexfmpe

MattSturgeon reviewed (dismissed):
With these changes, `override` is being added to the result, not the function.

I don't think this _particularly_ changes how lazily `f` is evaluated; rather it just moves the `override` attribute to the wrong place.

sternenseemann committed:
buildPython*: unconditionally add .override

To be able to splice, we can't fail to eval before returning an
attribute set. By checking for f ? override, we need to force f which
isn't always possible, whereas mirrorFunctionArgs serves as an
indirection which wraps the inner thunks in an attribute set that can
always be returned. Fortunately, there should be no case when f
evaluates successfully and does not have the override attribute, so we
can just remove the condition.

This fixes evaluation of buildPython* based packages in certain splicing
situations by being lazier. An example of this is
pkgsCross.ghcjs.buildPackages.nixpkgs-openjdk-updater on staging right
now (ff5098e).

MattSturgeon reviewed (commented):
This is a much better approach. Avoiding `(f ? override)` does indeed make `f` lazier.

> If `!(f ? override)` accessing `.override` of the returned function will[...]

We shouldn't need to worry about that edge-case, because `makeOverridablePythonPackage` is a private function that is always applied with a `callPackage` result, so `? override` will always be true.

With this in mind, the commit message could be shortened and put less emphasis on "we can't do (f?override)" and more emphasis on "we don't actually need to do it anyway"?

(I edited the PR description to match your revised approach)

SGTM overall, thanks!

sternenseemann commented:
> We shouldn't need to worry about that edge-case

Done.

MattSturgeon reviewed (approved):
> An example of this is `pkgsCross.ghcjs.buildPackages.nixpkgs-openjdk-updater` on staging right now (ff5098e).

Is this still accurate? https://github.com/NixOS/nixpkgs/pull/366593#issuecomment-3445062723 seems to contradict.

Either way, this is just nitpicking the commit wording; the change itself is good regardless of whether it addresses a currently-present issue.

---

LGTM, thanks!
